[BUG] test_arrays_zip failures on nightly

**Describe the bug**
```
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Byte)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304419.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Short)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304598.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Integer)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304777.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Long)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304956.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Float)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305135.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Double)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305314.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(String)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305493.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Boolean)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305672.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Date)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305851.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Timestamp)] - py4j.protocol.Py4JJavaError: An error occurred while calling o306038.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Null)] - py4j.protocol.Py4JJavaError: An error occurred while calling o306217.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(7,3))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306396.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(12,2))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306575.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(20,2))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306754.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Array(Short))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306933.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Array(String))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307112.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Struct(['child0', Byte],['child1', String],['child2', Float]))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307291.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Map(String(not_null),String))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307470.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Binary)] - py4j.protocol.Py4JJavaError: An error occurred while calling o307649.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip_corner_cases - py4j.protocol.Py4JJavaError: An error occurred while calling o307828.collec...
[2023-01-06T11:40:23.946Z] = 20 failed, 15074 passed, 830 skipped, 572 xfailed, 235 xpassed, 911 warnings in 6812.06s (1:53:32) =
```

Errors:
```
[2023-01-06T11:40:23.930Z] E                   	at java.lang.Thread.run(Thread.java:750)

[2023-01-06T11:40:23.930Z] E                   Caused by: java.lang.IllegalArgumentException: All columns must have the same number of rows

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnView.makeStructView(ColumnView.java:3440)

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnView.makeStructView(ColumnView.java:3458)

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnVector.makeStruct(ColumnVector.java:395)

[2023-01-06T11:40:23.930Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$zipArrays$2(collectionOperations.scala:825)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$zipArrays$1(collectionOperations.scala:824)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.zipArrays(collectionOperations.scala:823)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$columnarEval$5(collectionOperations.scala:782)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.columnarEval(collectionOperations.scala:732)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:109)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)

```

--- a/integration_tests/src/main/python/array_test.py
+++ b/integration_tests/src/main/python/array_test.py
@@ -417,7 +417,6 @@ array_zips_gen = array_gens_sample + [ArrayGen(map_string_string_gen[0], max_len
 
 
 @pytest.mark.parametrize('data_gen', array_zips_gen, ids=idfn)
-@pytest.mark.xfail(reason="https://github.com/NVIDIA/spark-rapids/issues/7469")
 def test_arrays_zip(data_gen):
     gen = StructGen(
         [('a', data_gen), ('b', data_gen), ('c', data_gen), ('d', data_gen)], nullable=False)
@@ -430,7 +429,6 @@ def test_arrays_zip(data_gen):
     )
 
 
-@pytest.mark.xfail(reason="https://github.com/NVIDIA/spark-rapids/issues/7469")
 def test_arrays_zip_corner_cases():
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark: unary_op_df(spark, ArrayGen(int_gen), length=100).selectExpr(

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021-2022, NVIDIA CORPORATION.
+ * Copyright (c) 2021-2023, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
@@ -729,61 +729,65 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
       GpuExpressionsUtils.columnarEvalToColumn(expr, batch).getBase
     }
 
-    withResource(inputs) { inputs =>
+    val cleanedInputs = withResource(inputs) { inputs =>
+      normalizeNulls(inputs)
+    }
 
-      // Compute max size of input arrays for each row
-      //
-      // input1: [[A, B, C], [D, E], [F], [G]]
-      // input2: [[a, b], [c, d, e], null, [f, g]]
-      // max array size: [3, 3, 0, 2]
-      val maxArraySize = computeMaxArraySize(inputs)
+    val padded = withResource(cleanedInputs) { cleanedInputs =>
+      padArraysToMaxLength(cleanedInputs)
+    }
 
-      // Generate offsets from max array sizes
-      //
-      // [3, 3, 0, 2] => [0, 3, 6, 6, 8]
-      val offsets = maxArraySize.generateListOffsets()
-
-      // Generate sequence indices for gathering children of input arrays
-      //
-      // [3, 3, 0, 2] => [[0, 1, 2], [0, 1, 2], [], [0, 1]]
-      val seqIndices = closeOnExcept(offsets) { _ =>
-        withResource(maxArraySize) { _ =>
-          generateSeqIndices(maxArraySize)
-        }
-      }
-
-      // Perform segment gather on input columns with indices covering each element
-      //
-      // input1: [[A, B, C], [D, E], [F], [G]]
-      // input2: [[a, b], [c, d, e], null, [f, g]]
-      // indices: [[0, 1, 2], [0, 1, 2], [], [0, 1]]
-      // children1: [A, B, C, D, E, null, G, null]
-      // children2: [a, b, null, c, d, e, f, g]
-      val gatheredChildren = closeOnExcept(offsets) { _ =>
-        withResource(seqIndices) { _ =>
-          inputs.safeMap { cv =>
-            withResource(cv.segmentedGather(seqIndices)) { gathered =>
-              gathered.getChildColumnView(0).copyToColumnVector()
-            }
-          }
-        }
-      }
-
-      // Zip gathered children along with a union offset in the form of List of Struct
-      //
-      // validity mask: [true, true, false, true]
-      // offsets: [0, 3, 6, 6, 8]
-      // children1: [A, B, C, D, E, null, G, null]
-      // children2: [a, b, null, c, d, e, f, g]
-      // zipped: [[{A, a}, {B, b}, {C, null}],
-      //          [{D, c}, {E, d}, {null, e}],
-      //          null,
-      //          [{G, f}, {null, g}]]
-      closeOnExcept(zipArrays(offsets, gatheredChildren, inputs)) { ret =>
+    withResource(padded) { _ =>
+      closeOnExcept(zipArrays(padded)) { ret =>
         GpuColumnVector.from(ret, dataType)
       }
     }
   }
+
+  /**
+   * Segmented gather in CUDF produces a NULL output for a NULL input. But we need to produce
+   * child columns that we can put together in a struct. This requires them all to have the
+   * same length. To make this work we need to make sure all of the inputs have nulls in the
+   * same places at the top level. That way when we gather things we get the same output
+   * size for all of the children of the input columns. The result of this will have the same
+   * validity for all top level columns, but possibly different offsets.
+   */
+  private def normalizeNulls(inputs: Seq[cudf.ColumnVector]): Seq[ColumnVector] = {
+    // First lest figure out if there are any nulls at all, because if there are not we don't
+    // need to do anything.
+    if (inputs.exists(_.hasNulls)) {
+      var nullOutput = inputs.head.isNull
+      try {
+        inputs.slice(1, inputs.length).foreach { cv =>
+          val combinedIsNull = withResource(cv.isNull) { tmpIsNull =>
+            tmpIsNull.or(nullOutput)
+          }
+          closeOnExcept(combinedIsNull) { _ =>
+            nullOutput.close()
+            nullOutput = combinedIsNull
+          }
+        }
+
+        // input1: [[A, B, C], [D, E], [F], [G]]
+        // input2: [[a, b], [c, d, e], null, [f, g]]
+        // combinedIsNull, false, false, true, false
+
+        // output1: [[A, B, C], [D, E], null, [G]]
+        // output2: [[a, b], [c, d, e], null, [f, g]]
+
+        inputs.zip(children).safeMap { case (cv, child) =>
+          withResource(GpuScalar.from(null, child.dataType)) { nullArray =>
+            nullOutput.ifElse(nullArray, cv)
+          }
+        }
+      } finally {
+        nullOutput.close()
+      }
+    } else {
+      inputs.map(_.incRefCount())
+    }
+  }
+
 
   private def computeMaxArraySize(inputs: Seq[ColumnVector]): ColumnVector = {
     // Compute array sizes of input arrays
@@ -809,28 +813,65 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
 
   private def generateSeqIndices(maxArraySize: ColumnVector): ColumnVector = {
     withResource(GpuScalar.from(0, IntegerType)) { s =>
-      val zeroCV = GpuColumnVector.from(s, maxArraySize.getRowCount.toInt, IntegerType)
-      withResource(zeroCV.getBase) { zero =>
+      withResource(cudf.ColumnVector.fromScalar(s, maxArraySize.getRowCount.toInt)) { zero =>
         ColumnVector.sequence(zero, maxArraySize)
       }
     }
   }
 
-  private def zipArrays(offsets: ColumnVector,
-                        children: Seq[ColumnVector],
-                        inputs: Seq[ColumnVector]) : ColumnVector = {
-    // Construct List of Struct with list offsets and struct fields
-    val zipped = withResource(offsets) { _ =>
-      withResource(children) { _ =>
-        withResource(ColumnVector.makeStruct(children: _*)) { struct =>
-          struct.makeListFromOffsets(offsets.getRowCount - 1, offsets)
-        }
+  /**
+   * Do a segmented gather on the inputs so that they are padded with nulls to make sure each LIST
+   * on the same row, at the top level, has the same length. The columns returned should have the
+   * same offsets and the same validity. This assumes that the validity on the inputs all match.
+   */
+  private def padArraysToMaxLength(inputs: Seq[ColumnVector]): Seq[ColumnVector] = {
+    // Compute max size of input arrays for each row, this is to know how we need to pad things.
+    //
+    // input1: [[A, B, C], [D, E], null, [G]]
+    // input2: [[a, b], [c, d, e], null, [f, g]]
+    // max array size: [3, 3, 0, 2]
+    val seqIndices = withResource(computeMaxArraySize(inputs)) { maxArraySize =>
+      // Generate sequence indices for gathering children of input arrays
+      //
+      // [3, 3, 0, 2] => [[0, 1, 2], [0, 1, 2], [], [0, 1]]
+      generateSeqIndices(maxArraySize)
+    }
+
+    // Perform segment gather on input columns with indices covering each element
+    //
+    // input1: [[A, B, C], [D, E], null, [G]]
+    // input2: [[a, b], [c, d, e], null, [f, g]]
+    // indices: [[0, 1, 2], [0, 1, 2], [], [0, 1]]
+    // output1: [[A, B, C], [D, E, null], null, [G, null]]
+    // output2: [[a, b, null], [c, d, e], null, [f, g]]
+    withResource(seqIndices) { _ =>
+      inputs.safeMap { cv =>
+        cv.segmentedGather(seqIndices)
       }
     }
-    // Set the validity of result with the joint validity of input columns, which means
-    // the output record will be null if any of input records is null.
-    withResource(zipped) { _ =>
-      zipped.mergeAndSetValidity(BinaryOp.BITWISE_AND, inputs: _*)
+  }
+
+  /**
+   * This turns LIST[X], LIST[Y], ... into a LIST[STRUCT[X, Y, ...]] but requires that
+   * the input LIST columns all have the same validity and offsets.
+   */
+  private def zipArrays(padded: Seq[ColumnVector]) : ColumnVector = {
+    // Get the data column from the children, without any offsets
+    val children = padded.safeMap(_.getChildColumnView(0))
+    // Put them into a struct column view
+    withResource(ColumnView.makeStructView(children: _*)) { structView =>
+      // Then make the struct a list using the input's offsets and validity
+      // in the cheapest way possible.
+      val proto = padded.head
+      withResource(proto.getValid) { valid =>
+        withResource(proto.getOffsets) { offsets =>
+          withResource(new ColumnView(DType.LIST, proto.getRowCount,
+            java.util.Optional.of[java.lang.Long](proto.getNullCount),
+            valid, offsets, Array(structView))) { retView =>
+            retView.copyToColumnVector()
+          }
+        }
+      }
     }
   }
 }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala
@@ -753,12 +753,12 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
    * validity for all top level columns, but possibly different offsets.
    */
   private def normalizeNulls(inputs: Seq[cudf.ColumnVector]): Seq[ColumnVector] = {
-    // First lest figure out if there are any nulls at all, because if there are not we don't
+    // First let's figure out if there are any nulls at all, because if there are not we don't
     // need to do anything.
     if (inputs.exists(_.hasNulls)) {
       var nullOutput = inputs.head.isNull
       try {
-        inputs.slice(1, inputs.length).foreach { cv =>
+        inputs.drop(1).foreach { cv =>
           val combinedIsNull = withResource(cv.isNull) { tmpIsNull =>
             tmpIsNull.or(nullOutput)
           }
@@ -813,7 +813,7 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
 
   private def generateSeqIndices(maxArraySize: ColumnVector): ColumnVector = {
     withResource(GpuScalar.from(0, IntegerType)) { s =>
-      withResource(cudf.ColumnVector.fromScalar(s, maxArraySize.getRowCount.toInt)) { zero =>
+      withResource(ColumnVector.fromScalar(s, maxArraySize.getRowCount.toInt)) { zero =>
         ColumnVector.sequence(zero, maxArraySize)
       }
     }
@@ -852,23 +852,25 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
   }
 
   /**
-   * This turns LIST[X], LIST[Y], ... into a LIST[STRUCT[X, Y, ...]] but requires that
+   * This turns LIST[X], LIST[Y], ... into a LIST[ STRUCT[X, Y, ...] ] but requires that
    * the input LIST columns all have the same validity and offsets.
    */
-  private def zipArrays(padded: Seq[ColumnVector]) : ColumnVector = {
+  private def zipArrays(padded: Seq[ColumnVector]): ColumnVector = {
     // Get the data column from the children, without any offsets
-    val children = padded.safeMap(_.getChildColumnView(0))
-    // Put them into a struct column view
-    withResource(ColumnView.makeStructView(children: _*)) { structView =>
-      // Then make the struct a list using the input's offsets and validity
-      // in the cheapest way possible.
-      val proto = padded.head
-      withResource(proto.getValid) { valid =>
-        withResource(proto.getOffsets) { offsets =>
-          withResource(new ColumnView(DType.LIST, proto.getRowCount,
-            java.util.Optional.of[java.lang.Long](proto.getNullCount),
-            valid, offsets, Array(structView))) { retView =>
-            retView.copyToColumnVector()
+    withResource(padded.safeMap(_.getChildColumnView(0))) { children =>
+      // Put them into a struct column view
+      withResource(ColumnView.makeStructView(children: _*)) { structView =>
+        // Make the struct a list using the input's offsets and validity
+        // in the cheapest way possible.
+        val proto = padded.head
+        withResource(proto.getValid) { valid =>
+          withResource(proto.getOffsets) { offsets =>
+            withResource(new ColumnView(DType.LIST, proto.getRowCount,
+              java.util.Optional.of[java.lang.Long](proto.getNullCount),
+              valid, offsets, Array(structView))) { retView =>
+              // Finally copy the result out to a ColumnVector so we can return it
+              retView.copyToColumnVector()
+            }
           }
         }
       }

